File: src/Transpiler.Backends/Runtime/numeric.py
# Explicit binary32 evaluation/storage and bounded little-endian CLI field data.
class CliDataHandle:
    __slots__ = ('data',)
    def __init__(self, data): self.data = data


class NumericRuntime(ValueRuntime):
    @staticmethod
    def f32(value):
        try: return _struct.unpack('<f', _struct.pack('<f', float(value)))[0]
        except OverflowError: return math.copysign(float('inf'), value)

    def coerce(self, value, type_name):
        return self.f32(value) if type_name == 'System.Single' else super().coerce(value, type_name)

    def binary(self, op, a, b, kind):
        return self.f32(super().binary(op, a, b, 'f')) if kind == 'f4' else super().binary(op, a, b, kind)

    def unary(self, op, value, kind):
        return self.f32(super().unary(op, value, 'f')) if kind == 'f4' else super().unary(op, value, kind)

    def convert(self, op, value, source):
        if op in ('conv.r4', 'conv.r4.from-unsigned'):
            if source in ('i4', 'i8'):
                integer = self.bits(value, 32 if source == 'i4' else 64, op.endswith('unsigned'))
                negative = integer < 0; magnitude = abs(integer)
                shift = max(0, magnitude.bit_length() - 24)
                rounded = magnitude >> shift
                if shift:
                    remainder = magnitude - (rounded << shift); halfway = 1 << (shift - 1)
                    if remainder > halfway or remainder == halfway and rounded & 1: rounded += 1
                result = float(rounded) * (2 ** shift)
                return -result if negative else result
            return self.f32(value)
        return super().convert(op, value, 'f' if source == 'f4' else source)

    def finite(self, value):
        if not math.isfinite(value): self.fail('System.ArithmeticException', 'Non-finite floating-point value.')
        return value

    def semantic_hash(self, value, type_name, dispatch=True):
        value = self.dereference(value)
        if type_name == 'System.Single' and not isinstance(value, CliBox):
            if value == 0: return 0
            if math.isnan(value): return 2139095040
            return _struct.unpack('<i', _struct.pack('<f', value))[0]
        return super().semantic_hash(value, type_name, dispatch)

    def format(self, value, type_name='System.Object'):
        value = self.dereference(value)
        if isinstance(value, CliBox): return self.format(value.value, value.type)
        if type_name != 'System.Single': return super().format(value, type_name)
        value = self.f32(value)
        if not math.isfinite(value) or value == 0: return self.double_text(value)
        # Shortest decimal that round-trips through binary32; invariant general-format exponent thresholds.
        for precision in range(1, 10):
            text = format(value, '.' + str(precision) + 'g')
            if self.f32(float(text)) == value: break
        negative = text.startswith('-'); mantissa, _, power = text.lstrip('-').lower().partition('e')
        whole, _, fraction = mantissa.partition('.')
        exponent = int(power or '0') + len(whole) - 1
        if whole == '0': exponent = int(power or '0') - (len(fraction) - len(fraction.lstrip('0'))) - 1
        digits = (whole + fraction).lstrip('0').rstrip('0') or '0'
        sign = '-' if negative else ''
        if exponent < -4 or exponent >= 9:
            return sign + digits[0] + ('.' + digits[1:] if len(digits) > 1 else '') + 'E' + ('+' if exponent >= 0 else '-') + str(abs(exponent)).zfill(2)
        position = exponent + 1
        if position <= 0: return sign + '0.' + '0' * -position + digits
        if position >= len(digits): return sign + digits + '0' * (position - len(digits))
        return sign + digits[:position] + '.' + digits[position:]

    def data_handle(self, key):
        data = self.meta['fields'][key]['data']
        if data is None: raise RuntimeError('Missing validated field data')
        return CliDataHandle(bytes(data))

    def initialize_data(self, array, handle):
        self.nonnull(array)
        if not isinstance(array, CliArray) or not isinstance(handle, CliDataHandle): self.fail('System.ArgumentException', 'Invalid array initialization arguments.')
        element = self.meta['types'].get(array.element, {}).get('enumType') or array.element
        encodings = {'System.Boolean':'B','System.SByte':'b','System.Byte':'B','System.Char':'H',
                     'System.Int16':'h','System.UInt16':'H','System.Int32':'i','System.UInt32':'I',
                     'System.Int64':'q','System.UInt64':'Q','System.Single':'f','System.Double':'d'}
        code = encodings.get(element)
        if code is None: self.fail('System.ArgumentException', 'Array initialization requires primitive or enum elements.')
        size = _struct.calcsize('<' + code)
        if len(array.data) > len(handle.data) // size: self.fail('System.ArgumentException', 'Field data is shorter than the array.')
        for index in range(len(array.data)):
            array.data[index] = self.coerce(_struct.unpack_from('<' + code, handle.data, index * size)[0], element)

    def external(self, method, args):
        op = method['intrinsic']
        if op == 'array.initialize-data': self.initialize_data(args[0], args[1]); return None
        if op == 'bits.single-i4': return _struct.unpack('<i', _struct.pack('<f', args[0]))[0]
        if op == 'bits.i4-single': return _struct.unpack('<f', _struct.pack('<i', self.bits(args[0],32)))[0]
        if op == 'bits.double-i8': return _struct.unpack('<q', _struct.pack('<d', args[0]))[0]
        if op == 'bits.i8-double': return _struct.unpack('<d', _struct.pack('<q', self.bits(args[0],64)))[0]
        if op.startswith('float.'):
            value = args[0]; test = op[6:]
            return int({'IsFinite':lambda:math.isfinite(value),'IsNaN':lambda:math.isnan(value),
                        'IsInfinity':lambda:math.isinf(value),'IsPositiveInfinity':lambda:value == float('inf'),
                        'IsNegativeInfinity':lambda:value == -float('inf'),'IsNegative':lambda:math.copysign(1,value)<0}[test]())
        if op.startswith('mathf.'):
            name = op[6:]
            if name == 'copysign': return self.f32(math.copysign(args[0], args[1]))
            proxy = dict(method, intrinsic='math.' + name, params=['System.Double'] * len(args))
            return self.f32(super().external(proxy, args))
        return super().external(method, args)

File: src/Transpiler.Backends/Runtime/numeric.mjs
// Explicit binary32 evaluation/storage and bounded little-endian CLI field data.
class CliDataHandle { constructor(data) { this.data = data; } }
class NumericRuntime extends ValueRuntime {
    coerce(value, type) { return type === 'System.Single' ? Math.fround(value) : super.coerce(value, type); }
    binary(op, a, b, kind) { return kind === 'f4' ? Math.fround(super.binary(op, a, b, 'f')) : super.binary(op, a, b, kind); }
    unary(op, value, kind) { return kind === 'f4' ? Math.fround(super.unary(op, value, 'f')) : super.unary(op, value, kind); }
    convert(op, value, source) {
        if (op === 'conv.r4' || op === 'conv.r4.from-unsigned') {
            if (source === 'i4' || source === 'i8') {
                const integer=BigInt(this.bits(value,source==='i4'?32:64,op.endsWith('unsigned')));
                const negative=integer<0n, magnitude=negative?-integer:integer;
                const shift=Math.max(0,magnitude.toString(2).length-24);
                let rounded=magnitude>>BigInt(shift);
                if(shift) {
                    const remainder=magnitude-(rounded<<BigInt(shift)), halfway=1n<<BigInt(shift-1);
                    if(remainder>halfway||(remainder===halfway&&(rounded&1n)!==0n)) rounded++;
                }
                const result=Number(rounded)*2**shift;
                return negative?-result:result;
            }
            return Math.fround(value);
        }
        return super.convert(op, value, source === 'f4' ? 'f' : source);
    }
    finite(value) {
        if (!Number.isFinite(value)) this.fail('System.ArithmeticException', 'Non-finite floating-point value.');
        return value;
    }
    semantic_hash(value, type, dispatch = true) {
        value = this.dereference(value);
        if (type === 'System.Single' && !(value instanceof CliBox)) {
            if (value === 0) return 0;
            if (Number.isNaN(value)) return 2139095040;
            const view = new DataView(new ArrayBuffer(4)); view.setFloat32(0, value, true); return view.getInt32(0, true);
        }
        return super.semantic_hash(value, type, dispatch);
    }
    format(value, type = 'System.Object') {
        value = this.dereference(value);
        if (value instanceof CliBox) return this.format(value.value, value.type);
        if (type !== 'System.Single') return super.format(value, type);
        value = Math.fround(value);
        if (!Number.isFinite(value) || value === 0) return this.double_text(value);
        let text;
        for (let precision = 1; precision <= 9; precision++) {
            text = value.toPrecision(precision);
            if (Math.fround(Number(text)) === value) break;
        }
        const sign = text.startsWith('-') ? '-' : '';
        const [mantissa,power='0'] = text.replace(/^-/, '').toLowerCase().split('e');
        const [whole,fraction=''] = mantissa.split('.');
        let exponent = Number(power) + whole.length - 1;
        if (whole === '0') exponent = Number(power) - (fraction.length - fraction.replace(/^0+/, '').length) - 1;
        const digits = (whole+fraction).replace(/^0+/, '').replace(/0+$/, '') || '0';
        if (exponent < -4 || exponent >= 9) return sign + digits[0] + (digits.length>1 ? '.'+digits.slice(1) : '') + 'E' + (exponent>=0 ? '+' : '-') + String(Math.abs(exponent)).padStart(2,'0');
        const position = exponent+1;
        if(position<=0) return sign+'0.'+'0'.repeat(-position)+digits;
        if(position>=digits.length) return sign+digits+'0'.repeat(position-digits.length);
        return sign+digits.slice(0,position)+'.'+digits.slice(position);
    }
    data_handle(key) {
        const data = this.meta.fields[key].data;
        if (data === null) throw new Error('Missing validated field data');
        return new CliDataHandle(new Uint8Array(data));
    }
    initialize_data(array, handle) {
        this.nonnull(array);
        if (!(array instanceof CliArray) || !(handle instanceof CliDataHandle)) this.fail('System.ArgumentException','Invalid array initialization arguments.');
        const element = this.meta.types[array.element]?.enumType ?? array.element;
        const formats = {'System.Boolean':[1,'getUint8'],'System.SByte':[1,'getInt8'],'System.Byte':[1,'getUint8'],
            'System.Char':[2,'getUint16'],'System.Int16':[2,'getInt16'],'System.UInt16':[2,'getUint16'],
            'System.Int32':[4,'getInt32'],'System.UInt32':[4,'getUint32'],'System.Int64':[8,'getBigInt64'],
            'System.UInt64':[8,'getBigUint64'],'System.Single':[4,'getFloat32'],'System.Double':[8,'getFloat64']};
        const format = formats[element];
        if (!format) this.fail('System.ArgumentException','Array initialization requires primitive or enum elements.');
        const [size,reader] = format;
        if (array.data.length > Math.floor(handle.data.length/size)) this.fail('System.ArgumentException','Field data is shorter than the array.');
        const view = new DataView(handle.data.buffer,handle.data.byteOffset,handle.data.byteLength);
        for(let index=0; index<array.data.length; index++) array.data[index] = this.coerce(view[reader](index*size,true),element);
    }
    external(method,args) {
        const op = method.intrinsic;
        if(op === 'array.initialize-data') { this.initialize_data(args[0],args[1]); return null; }
        if(op.startsWith('bits.')) {
            const view=new DataView(new ArrayBuffer(8));
            switch(op) {
                case 'bits.single-i4': view.setFloat32(0,args[0],true); return view.getInt32(0,true);
                case 'bits.i4-single': view.setInt32(0,args[0],true); return view.getFloat32(0,true);
                case 'bits.double-i8': view.setFloat64(0,args[0],true); return view.getBigInt64(0,true);
                case 'bits.i8-double': view.setBigInt64(0,args[0],true); return view.getFloat64(0,true);
                default: throw new Error('Unknown validated bit conversion');
            }
        }
        if(op.startsWith('float.')) {
            const value=args[0];
            switch(op.slice(6)) {
                case 'IsFinite': return Number(Number.isFinite(value));
                case 'IsNaN': return Number(Number.isNaN(value));
                case 'IsInfinity': return Number(value===Infinity||value===-Infinity);
                case 'IsPositiveInfinity': return Number(value===Infinity);
                case 'IsNegativeInfinity': return Number(value===-Infinity);
                case 'IsNegative': { const v=new DataView(new ArrayBuffer(8));v.setFloat64(0,value,true);return v.getUint32(4,true)>>>31; }
            }
        }
        if(op.startsWith('mathf.')) {
            const name=op.slice(6);
            if(name==='copysign') return Math.fround(args[1]<0||Object.is(args[1],-0) ? -Math.abs(args[0]) : Math.abs(args[0]));
            return Math.fround(super.external({...method,intrinsic:'math.'+name,params:args.map(_=>'System.Double')},args));
        }
        return super.external(method,args);
    }
}
